src/libappimage/desktop_integration/integrator/Integrator.cpp

```diff
@@ -0,0 +1,267 @@
+             */
+            struct Integrator::Priv {
+                core::AppImage appImage;
+                bf::path xdgDataHome;
+                std::string appImageId;
+                std::string vendorPrefix = "appimagekit";
+
+                DesktopEntry entry;
+                DesktopIntegrationResources resources;
+
+                Priv(const AppImage& appImage, const std::string& xdgDataHome) : appImage(appImage),
+                                                                                 xdgDataHome(xdgDataHome) {
+
+                    if (xdgDataHome.empty())
+                        Priv::xdgDataHome = XdgUtils::BaseDir::XdgDataHome();
+                }
+
+                /**
+                 * The AppImage Id is build from the MD5 hash sum of it's path.
+                 */
+                void buildAppImageId() {
+                    appImageId = appimage_get_md5(appImage.getPath().c_str()) ?: "";
+                }
+
+                void readResources() {
+                    try {
+                        ResourcesExtractor extractor(appImage);
+
+                        resources = extractor.extract();
+                        if (resources.desktopEntryPath.empty())
+                            throw DesktopIntegrationError("Missing Desktop File");
+
+                    } catch (const AppImageError& error) {
+                        throw DesktopIntegrationError("Unable to extract the desktop integration resources.");
+                    }
+                }
+
+                void loadDesktopEntry() {
+                    std::string desktopEntryData(resources.desktopEntryData.begin(),
+                                                 resources.desktopEntryData.end());
+
+                    std::stringstream desktopEntryDataStream(desktopEntryData);
+
+                    desktopEntryDataStream >> entry;
+
+                    // Check if the AppImage author requested that this AppImage should not be integrated
+                    auto integrationRequest = entry.get("Desktop Entry/X-AppImage-Integrate");
+                    boost::algorithm::erase_all(integrationRequest, " ");
+                    boost::algorithm::to_lower(integrationRequest);
+                    if (integrationRequest == "false")
+                        throw DesktopIntegrationError("The AppImage explicitly requested to not be integrated");
+                }
+
+                void deployDesktopEntry() {
+                    editDesktopEntry(entry, appImageId);
+
+                    // deploy desktop entry
+                    bf::path desktopEntryDeployPath = buildDesktopFilePath();
+
+                    // create dir
+                    create_directories(desktopEntryDeployPath.parent_path());
+
+                    // write file contents
+                    std::ofstream desktopEntryFile(desktopEntryDeployPath.string());
```

Comment:
It will throw an exception which will stop the integration process as expected. Maybe what we should add is a clean up in case of errors. But that's a topic for another issue #60